data_analysis/near-deduplication/minhash_deduplication_alt.py

```diff
@@ -5,697 +5,331 @@
+        """Source: `datasketch.lsh`"""
 
-    Returns
-    -------
-    float
-        The jaccard similarity between the two code snippets.
-
-    Examples
-    --------
-    >>> jaccard_similarity("a = 1", "a = 2")
-    0.3333333333333333
-    >>> jaccard_similarity("a = 1", "a = 1")
-    1.0
-    """
-    tokens1 = set([t for t in NON_ALPHA.split(code1) if t.strip()])
-    tokens2 = set([t for t in NON_ALPHA.split(code2) if t.strip()])
-    return len(tokens1 & tokens2) / max(1, len(tokens1 | tokens2))
+        def proba(s):
+            return 1 - (1 - (1 - s ** float(r)) ** float(b))
 
+        a, _ = integrate(proba, threshold, 1.0)
+        return a
 
-def find_duplicate_non_extremes(
-    records: Iterable | Dataset,
-    reference_records: Iterable | Dataset,
-    column: str,
-    threshold: float,
-) -> None:
-    """
-    This is a approximation of what has been used in other script.
+    min_error = float("inf")
+    opt = (0, 0)
+    for b in range(1, num_perm + 1):
+        max_r = int(num_perm / b)
+        for r in range(1, max_r + 1):
+            fp = false_positive_probability(threshold, b, r)
+            fn = false_negative_probability(threshold, b, r)
+            error = fp * false_positive_weight + fn * false_negative_weight
+            if error < min_error:
+                min_error = error
+                opt = (b, r)
+    return opt
 
-    This is slow in this implementation as parallelization requires a global variable
-    to hold the dataset to query, which is not implemented in this script.
-    """
-    g = nk.graph.Graph()
-    for record in tqdm(records, desc="Constructing graph..."):
-        for y in record["__neighbors__"]:
-            g.addEdge(record["__id__"], y, addMissing=True)
-
-    to_remove: Set[int] = set()
-    cc = nk.components.ConnectedComponents(g)
-    cc.run()
-    for component in tqdm(sorted(cc.getComponents(), key=len, reverse=True), desc="Iterating over components..."):
-        extremes: Set[int] = set()
-        # greedy clustering within each component
-        for element1 in tqdm(component, leave=False):
-            code1 = reference_records[element1][column]
-            for element2 in extremes:
-                code2 = reference_records[element2][column]
-                if jaccard_similarity(code1, code2) >= threshold:
-                    break
-            else:
-                extremes.add(element1)
-        to_remove.update([i for i in component if i not in extremes])
-
-    return to_remove
+
+class UnionFind:
+    def __init__(self):
+        self.parent: Dict[int, int] = {}
+
+    def find(self, x):
+        if x not in self.parent:
+            self.parent[x] = x
+        if self.parent[x] != x:
+            self.parent[x] = self.find(self.parent[x])
+        return self.parent[x]
+
+    def union(self, x, y):
+        px = self.find(x)
+        py = self.find(y)
+        self.parent[px] = self.parent[py] = min(px, py)
```

Comment:
Not sure if it will give any real improvements, but maybe use rank-optimization (https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Union_by_rank) for choosing who is new parent?

Reply:
That's a good point to think about, thanks. But for this PR, which is to provide a reference point along with the SantaCoder paper that has been published, I will skip any functional change for integrity and reproducibility.